### PR TITLE
Add log line to `ConstantLit::fullUnresolvedPath`

### PR DESCRIPTION
**Pull request by jez**
### Motivation

This crash has been happening in Stripe's codebase for quite some time.
We've fixed a couple of the problems but clearly not enough, so I'd like
to figure out where the rest of them are coming from.

### Test plan

If it were possible to test this, I wouldn't have to make this PR.

### CODE REVIEW

File: ast/Trees.cc [hunk 1]
```diff
@@ -304,7 +304,11 @@ optional<pair<core::SymbolRef, vector<core::NameRef>>> ConstantLit::fullUnresolv
     if (this->symbol != core::Symbols::StubModule()) {
         return nullopt;
     }
-    ENFORCE(this->resolutionScopes != nullptr && !this->resolutionScopes->empty(), "loc={}", this->loc.showRaw(ctx));
+    if (this->resolutionScopes == nullptr || !this->resolutionScopes->empty()) [[unlikely]] {
+        ENFORCE(false);
+        fatalLogger->error(R"(msg="Bad fullUnresolvedPath" path="{}")");
+        fatalLogger->error("source=\"{}\"", absl::CEscape(ctx.file.data(ctx).source()));
```
Comment (neilparikh) [inline ast/Trees.cc:310]: You need include `absl/strings/escaping.h`

File: ast/Trees.cc [hunk 2]
```diff
@@ -304,7 +304,11 @@ optional<pair<core::SymbolRef, vector<core::NameRef>>> ConstantLit::fullUnresolv
     if (this->symbol != core::Symbols::StubModule()) {
         return nullopt;
     }
-    ENFORCE(this->resolutionScopes != nullptr && !this->resolutionScopes->empty(), "loc={}", this->loc.showRaw(ctx));
+    if (this->resolutionScopes == nullptr || !this->resolutionScopes->empty()) [[unlikely]] {
```
Comment (neilparikh) [inline ast/Trees.cc:307]: Should the condition in the `if` be `this->resolutionScopes == nullptr || this->resolutionScopes->empty()` (no `!` in front of `this->resolutionScopes->empty()`)?

Also, does [[unlikely]] do anything in C++17?

Comment (jez) [reply ast/Trees.cc:307]: good catch on the condition, thanks.

> Also, does [[unlikely]] do anything in C++17?

My understanding is that clang supports this despite it not being standardized until C++20

File: ast/Trees.cc [hunk 3]
```diff
@@ -304,7 +305,11 @@ optional<pair<core::SymbolRef, vector<core::NameRef>>> ConstantLit::fullUnresolv
     if (this->symbol != core::Symbols::StubModule()) {
         return nullopt;
     }
-    ENFORCE(this->resolutionScopes != nullptr && !this->resolutionScopes->empty(), "loc={}", this->loc.showRaw(ctx));
+    if (this->resolutionScopes == nullptr && this->resolutionScopes->empty()) [[unlikely]] {
```
Comment (neilparikh) [inline ast/Trees.cc:308]: I think the old `||` was correct, you can't call `empty()` on `nullptr`.
```suggestion
    if (this->resolutionScopes == nullptr || this->resolutionScopes->empty()) [[unlikely]] {
```

**Review by neilparikh — APPROVED** [review]

File: ast/Trees.cc (hunk 3 above)
Comment (jez) [reply ast/Trees.cc:308]: #8274

**Pull request merged**